### PR TITLE
[pipelining] Add schedule send/recv pass

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131762
* #130860
* #130488
* #132369
* __->__ #130378
* #129810

Inserts send/recv ops where needed in a compute-only pipeline schedule.

Any F or B action will require a recv op for its input and a send op
for its output, except for at the ends of the pipeline.

To avoid hangs caused by mixed-up orderings of sends/recvs across ranks,
we pick one compute action at a time and insert both its send op (on
that rank's schedule), and the matching recv op for the recipient stage
(on the schedule for the rank for that stage).

TODO
Currently ignores a couple of edge cases
- ignores batching (which is an optimization)
- ignores cases where a stage sends to anotehr stage on the same rank,
  and should skip the send/recv and directly access memory

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -823,6 +854,83 @@ def _reshard(stage_index: int):
...
+
+            if not _ready_to_schedule(action, comm_actions[rank]):
+                continue
+
+            if action is not None:
+                comm_actions[rank].append(action)
+                if _has_comms(action):
+                    send, recv = _get_comms(action)
+                    # TODO we can avoid send/recv if the 2 stages are on the same rank.
+                    # should we avoid that in the runtime or here?
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:922]: i think for the same rank sends, the logic will need to be handled in the runtime and we'll need to surface a different API at the pipeline stage level

File: torch/distributed/pipelining/schedules.py [hunk 2]
```diff
@@ -101,7 +132,7 @@ def from_str(str):
         elif str == "" or str.isspace():
             return None
         raise RuntimeError(
-            f"Invalid action string: {str}, should be formatted as [stage][action type][microbatch] e.g. 2F0"
+            f"Invalid action string: {str}, should be formatted as [stage][action type][(microbatch)] e.g. 2F0"
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:135]: Optional

File: torch/distributed/pipelining/schedules.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:922]: well, i agree thats _an_ option. But it should also be possible to do it at the IR level.  

(1) We can have an action type for SHMEM_SEND and SHMEM_RECV or something, and then the runtime can act on those actions.  

(2) Or we can delete the send/recv from the schedule IR, and the runtime would have to keep track of same-stages and do an extra step.  (This feels a little more clunky to me actually).

(3) Or we can keep the send/recv and the runtime can refuse to execute the send op if it sends to the same rank as it recvs from, and then do the same path as 2, but again it feels a bit clunky

what other options do you see?

**Review by H-Huang — APPROVED** [review]
LGTM

**Pull request closed without merging**